### ISSUE TITLE
PyPI default endpoint shutdown on July 3rd, 2017

### ISSUE THREAD

**Issue by lmazuel** [travis-ci/dpl#655]
See this message of @dstufft on distutils-sig:
https://mail.python.org/pipermail/distutils-sig/2017-June/030766.html

> As most people are aware, there has been an effort under way to rewrite PyPI in
> order to solve a lot of long standing problems. For those who aren't aware, that
> is currently available at https://pypi.org/ and it uses the same database that
> "Legacy" PyPI does, so the two are essentially just different views over the
> same data.
> 
> For a while now, Python, setuptools, and twine have all defaulted to using
> this new code base for uploading artifacts to PyPI. Now that we've gotten some
> testing of that code base, the infrastructure team and myself feel comfortable
> directing everyone to using the new endpoint and we're planning on shutting
> down uploads to Legacy PyPI.
> 
> If you're using the latest versions of Python, setuptools, or twine and you
> have not put an explicit URL in your ~/.pypirc file, then there's nothing you
> should need to do. If you are not, then you should ideally upgrade to the latest
> version of whatever tool you're using to upload (the preferred tool is twine)
> and edit your ~/.pypirc so that it removes any explicit mention of an URL. Thus
> it should look something like:
> 
>     [distutils]
>     index-servers =
>         pypi
> 
>     [pypi]
>     username:yourusername
>     password:yourpassword
> 
> 
> If for some reason you're not able to update to the newest version of your
> upload tool, then you can configure it to upload to the new code base by
> switching the URL to use https://upload.pypi.org/legacy/ instead of
> https://pypi.python.org/pypi. Thus your ~/.pypirc would then become:
> 
>     [distutils]
>     index-servers =
>         pypi
> 
>     [pypi]
>     repository:https://upload.pypi.org/legacy/
>     username:yourusername
>     password:yourpassword
> 
> 
> For those of you who are using TestPyPI, that will also be affected, and the
> required URL for the new upload endpoint for TestPyPI is
> https://test.pypi.org/legacy/.
> 
> We plan to disable uploads to legacy PyPI on July 3rd, 2017 so any configuration
> change will need to be made before that date. In addition, we plan to have a
> "brownout" on June 29th where we will shut the legacy endpoint down for that
> day.
> 
> For TestPyPI the change to disable uploads to legacy will be made in the next
> couple of days, likely this weekend.
> 
> As part of the error message that users will get when attempting to upload to
> legacy PyPI, we will include a link to a page that details how to ensure that
> they are using the new code base and not the legacy code base.
> 
> Thanks everyone!

Please update the default endpoint to avoid a massive PyPI deploy failure for Travis projects on July 3rd 2017 :(

**Comment by BanzaiMan** [comment]
@dstufft Do we just need to remove [this `repository` key](https://github.com/travis-ci/dpl/blob/e0ca95ed0b6d19baec2ff2d5e54a47920002d025/lib/dpl/provider/pypi.rb#L53)? Our `~/.pypirc` is written by this method: https://github.com/travis-ci/dpl/blob/e0ca95ed0b6d19baec2ff2d5e54a47920002d025/lib/dpl/provider/pypi.rb#L72-L79.

**Comment by dstufft** [comment]
@BanzaiMan Or default to ``https://upload.pypi.org/legacy/`` instead. I don't know ruby well enough to follow that code 100%.

**Comment by BanzaiMan** [comment]
Defaulting to `https://upload.pypi.org/legacy` is suggested in https://github.com/travis-ci/dpl/pull/526. (But has not received the small fix for the failing spec.)

The `pypi-no-legacy` branch does what's described above. Please test it, so that it can be released before the July 3rd deadline.

```yaml
deploy:
  provider: pypi
  edge:
    branch: pypi-no-legacy
  ⋮ # rest
```

**Comment by lmazuel** [comment]
Hi @BanzaiMan 
I was about to test but read the code first. Does your commit means that "PYPI_SERVER" (as an environment variable and/or configuration in travis.yaml) can't be used anymore? If yes, it's a problem, since people can want to release on PyPI test and/or their own "PyPI like" server.
My initial thought was that just changing "DEFAULT_SERVER" at the top of the file was enough. Do I miss something?

**Comment by lmazuel** [comment]
Hi @BanzaiMan , any update here? Time is short unfortunately :(

**Comment by lmazuel** [comment]
For people who find this issue before the fix, you just have to update PYPI_SERVER in your Travis settings:
![image](https://user-images.githubusercontent.com/1050156/27715443-02ba43c8-5cec-11e7-8753-3dd8e5cde0a2.png)

**Comment by BanzaiMan** [comment]
I apologize for the lack of updates. It is unfortunately not clear to me what is needed to prepare for the upcoming switch. I do not have a good repository to test, either.

My proposed fix simply removes the default PyPI server value https://github.com/travis-ci/dpl/compare/master...pypi-no-legacy. This is my interpretation of what @dstufft's email said. We should be installing twine https://github.com/travis-ci/dpl/blob/7b63975cf1c1d178cf290510c559ba63f2c47824/lib/dpl/provider/pypi.rb#L45, so I understood that to mean the switch to https://upload.pypi.org/legacy/ was unnecessary. I do admit that there is an outstanding issue https://github.com/travis-ci/travis-ci/issues/7735 that indicates there is a problem with our current twine installation. There is a proposed fix for that https://github.com/travis-ci/dpl/pull/630, but I have not received any report on how effective it is.

**Comment by BanzaiMan** [comment]
Note that there is no branch on this repository that does both (switching off PyPI server default and installing twine with user privilege).

**Comment by lmazuel** [comment]
Yes, we should remove pypiserver **if publishing to pypi**. You might want to publish to pypi test, or a personal pypi server that respects the pypi protocol ([example](https://pypi.python.org/pypi/pypiserver)).
So either you change the default value (and you specify it anytime and don't let Twine does it), or if "pypi server" is not set by the user, you don't write it in the configuration (and let twine do the job).
But it's important than PYPI_SERVER can still be configured.

**Comment by lmazuel** [comment]
TL;DR; 
- Possible fix 1: Update DEFAULT_SERVER to `https://upload.pypi.org/legacy/`
- Possible fix 2: If pypi_server or PYPI_SERVER is not set, don't write "repository" in the file

**Comment by BanzaiMan** [comment]
@dstufft Is there a plan to shut down https://upload.pypi.org/legacy/ at some point in the future, and if so, when?

**Comment by dstufft** [comment]
No plans at this point in time.

**Comment by BanzaiMan** [comment]
Very well, then. I'll switch the default, then release 1.8.39.

**Comment by BanzaiMan** [comment]
Could someone test it with:

```yaml
deploy:
  provider: pypi
  edge:
    branch: pypi-legacy
  ⋮ # rest
```

**Comment by jacquerie** [comment]
I can confirm that using that branch correctly created https://pypi.python.org/pypi/inspire-dojson/41.0.2 for us. But I have a question for you: when will that branch be removed?

**Comment by BanzaiMan** [comment]
@jacquerie I'll leave it there for a while, but not much after the 1.8.39 release (say 1 week?).

**Comment by BanzaiMan** [comment]
I've just pushed `dpl` 1.8.39, the only change from the previous 1.8.38 release is this issue.

As I've said earlier, `pypi-legacy` branch will be alive for about a week.

I'm closing this now.

**Comment by BanzaiMan** [comment]
@jacquerie I'm about to delete the `pypi-legacy` branch, just so you know.
